Feature/18 integrate coverage into travis

cryptomental opened the pull request:
I had lots of fun with this one.

Apart from the integrations I had to change hardcoded gas limit in 2_deploy_contracts.js during migration to accomodate for higher cost of the instrumented code. Perhaps a better estimate could be achieved by checking gas used in both cases on develop network and coverage network with instrumentation.

The two failing test cases during coverage run fail when sending half of the inital collateralToken supply, I am wondering whether this is a bug in the solidity-coverage. I had a brief look and there are a few things to watch for e.g. https://github.com/sc-forks/solidity-coverage/issues/118 

After the pull request is accepted I will likely integrate https://coveralls.io/features to get better overview on coverage statistics over time.

Passing build: https://travis-ci.org/cryptomental/MarketProtocol/builds/324991808

cryptomental commented:
There seems to be a problem with network not initialized yet after docker-compose up -d

{code}
Deploying MarketContractOraclize...
  ... 0x1becc89df86af7c511ce1ef60c010b223da9f89065d47fb61c7eebba8fd4f7d4
Error encountered, bailing. Network state unknown. Review successful transactions manually.
Error: VM Exception while processing transaction: revert
{code}

which disappears after a few seconds. I blindly increased sleep to 30 seconds in .travis.yml but a better solution must be found like described at https://docs.docker.com/compose/startup-order/

cryptomental commented:
No, that's actually something different, calling 'truffle test' or 'truffle migrate' the first time fails. I need to experiment a bit more before merging.

pelsasser commented:
@cryptomental - thanks for the hard work and updates.

cryptomental commented:
I increased sleep to 60 seconds and it passed. This is ugly. Perhaps containers on Travis take longer to initialize than my localhost (which makes sense). I will check if nicer solution can be found but this may be a bit time consuming.

cryptomental commented:
I carefully checked docker container logs when running in foreground mode.

The Oraclize Connector on eth bridges fully initializes after almost one minute:

```
eth-bridge_1           | [2018-01-04T13:15:57.053Z] INFO successfully deployed all contracts
eth-bridge_1           | [2018-01-04T13:15:57.056Z] INFO instance configuration file saved to /ethereum-bridge/config/instance/oracle_instance_20180104T131557.json
eth-bridge_1           | 
eth-bridge_1           | Please add this line to your contract constructor:
eth-bridge_1           | 
eth-bridge_1           | OAR = OraclizeAddrResolverI(0x6f485C8BF6fc43eA212E93BBF8ce046C7f1cb475);
eth-bridge_1           | 
eth-bridge_1           | [2018-01-04T13:15:57.063Z] WARN re-org block listen is disabled
eth-bridge_1           | [2018-01-04T13:15:57.063Z] INFO Listening @ 0xf7e3e47e06f1bddecb1b2f3a7f60b6b25fd2e233 (Oraclize Connector)
eth-bridge_1           | 
eth-bridge_1           | (Ctrl+C to exit)
eth-bridge_1           | 
truffle-coverage_1     | eth_getTransactionReceipt
eth-bridge-coverage_1  | [2018-01-04T13:15:57.280Z] INFO successfully deployed all contracts
eth-bridge-coverage_1  | [2018-01-04T13:15:57.283Z] INFO instance configuration file saved to /ethereum-bridge/config/instance/oracle_instance_20180104T131557.json
eth-bridge-coverage_1  | 
eth-bridge-coverage_1  | Please add this line to your contract constructor:
eth-bridge-coverage_1  | 
eth-bridge-coverage_1  | OAR = OraclizeAddrResolverI(0x6f485C8BF6fc43eA212E93BBF8ce046C7f1cb475);
eth-bridge-coverage_1  | 
eth-bridge-coverage_1  | [2018-01-04T13:15:57.289Z] WARN re-org block listen is disabled
eth-bridge-coverage_1  | [2018-01-04T13:15:57.289Z] INFO Listening @ 0xae5e351fc523dd203c7a5ec603bdb8fb1ed1d157 (Oraclize Connector)
```

It can be seen from the logs that once it is initializes, it writes instance configuration file to /ethereum-bridge/config/instance/ folder. I think the proper solution is to remove the dummy sleep in /travis.yml and execute a background script that will monitor the folder the config file is available. I will experiment a bit.

perfectmak commented:
@cryptomental I'm looking forward to the results of your experiments. I knew that sleep was flaky.

cryptomental commented:
I added a helper script https://github.com/cryptomental/ethereum-bridge-coverage/commit/ffe18dc1452e65f4991480eaa921c52313aab03b 

It takes on average 23-29 seconds on Travis to fully initialize Oraclize connector.

I also changed docker volume settings in docker-compose.yml to two named volumes and created tmpfs on build/node_modules folders as they were overlapping and I initially thought that this caused Travis failures.

The whole setup works correctly on my local workstation but for some reason eth bridge on Travis exits with 0 exit code, hence the failures occur. I am trying to find a root cause now.

pelsasser commented:
@cryptomental - thanks for the hard work!

cryptomental commented:
It looks like the problem is purely a timing issue. Truffle is not fully initialized yet, and eth-bridge tries to connect to truffle's port, then it exits. The previous setup works fine on Travis with two docker containers but fails with four. I will need to add wait-for mechanism, will have a look today evening if that helps.

cryptomental commented:
Woohoo problem solved! I believe the solution is clean and stable.

Pull request is ready for the review. Have a great weekend!

pelsasser commented on docker-compose.yml:
```diff
@@ -1,19 +1,55 @@
 
   eth-bridge:
-    image: perfectmak/ethereum-bridge:0.5.5
+    image: cryptomental/ethereum-bridge-coverage:0.5.6
```
@cryptomental - just to be sure - If I fork the ethereum bridge you modified into the MarketProtocol repo and change this line, we should be okay?

cryptomental replied:
The changes are small, see here https://github.com/perfectmak/ethereum-bridge/compare/master...cryptomental:master

After the PR we could merge the changes into MarketProtocol's ethereum bridge and push the docker image to docker hub from there, to have marketprotocol/ethereum-bridge in that line. Or try to push this upstream into the original eth bridge, but I am not sure that they will accept the increased gas limit change.

pelsasser commented:
This is awesome.  Will merge once I review, thanks again!
